### ISSUE TITLE
Unfreeze failed with redis in kubernetes

### ISSUE THREAD

**Issue by mingjie-li** [redisson/redisson#4356]
**Actual behavior**

We run our redis sentinel cluster in kubernetes with bitnami redis helm chart. 
After the slave redis pod deleted/recreated by Kubernetes, the pod ip of the new slave changed, but the pod hostname remain same. Redisson try to call slaveUp with old ip addr for the new slave pod, and we can see exception in the log as below :
```
13:04:02.760 [redisson-netty-2-17] ERROR org.redisson.connection.balancer.LoadBalancerManager - Unable to unfreeze entry: [freeSubscribeConnectionsAmount=0, freeSubscribeConnectionsCounter=value:64:queue:0, freeConnectionsAmount=0, freeConnectionsCounter=value:92:queue:0, freezeReason=SYSTEM, client=[addr=redis://<old.pod.ip.here>:6379], nodeType=SLAVE, firstFail=1654880642760]
java.util.concurrent.CompletionException: org.redisson.client.RedisConnectionException: Unable to connect to Redis server: <old.pod.ip.here>/<old.pod.ip.here>:6379
	at java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:292) ~[?:1.8.0_191]
	at java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:308) ~[?:1.8.0_191]
	at java.util.concurrent.CompletableFuture.biRelay(CompletableFuture.java:1284) ~[?:1.8.0_191]
	at java.util.concurrent.CompletableFuture$BiRelay.tryFire(CompletableFuture.java:1270) ~[?:1.8.0_191]
	at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:474) ~[?:1.8.0_191]
	at java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:1977) ~[?:1.8.0_191]
	at org.redisson.connection.pool.ConnectionPool.lambda$createConnection$1(ConnectionPool.java:151) ~[redisson-3.17.3.jar:3.17.3]
	at java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:760) ~[?:1.8.0_191]
	at java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:736) ~[?:1.8.0_191]
	at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:474) ~[?:1.8.0_191]
	at java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:1977) ~[?:1.8.0_191]
	at org.redisson.connection.pool.ConnectionPool.promiseFailure(ConnectionPool.java:307) ~[redisson-3.17.3.jar:3.17.3]
	at org.redisson.connection.pool.ConnectionPool.lambda$createConnection$6(ConnectionPool.java:273) ~[redisson-3.17.3.jar:3.17.3]
	at java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:760) [?:1.8.0_191]
	at java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:736) [?:1.8.0_191]
	at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:474) [?:1.8.0_191]
	at java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:1977) [?:1.8.0_191]
	at org.redisson.client.RedisClient$1$2.run(RedisClient.java:235) [redisson-3.17.3.jar:3.17.3]
	at io.netty.util.concurrent.AbstractEventExecutor.runTask(AbstractEventExecutor.java:174) [netty-common-4.1.77.Final.jar:4.1.77.Final]
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:167) [netty-common-4.1.77.Final.jar:4.1.77.Final]
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:470) [netty-common-4.1.77.Final.jar:4.1.77.Final]
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:503) [netty-transport-4.1.77.Final.jar:4.1.77.Final]
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:995) [netty-common-4.1.77.Final.jar:4.1.77.Final]
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74) [netty-common-4.1.77.Final.jar:4.1.77.Final]
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30) [netty-common-4.1.77.Final.jar:4.1.77.Final]
	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_191]
Caused by: org.redisson.client.RedisConnectionException: Unable to connect to Redis server: <old.pod.ip.here>/<old.pod.ip.here>:6379
	at org.redisson.connection.pool.ConnectionPool.lambda$createConnection$1(ConnectionPool.java:150) ~[redisson-3.17.3.jar:3.17.3]
	... 19 more
Caused by: java.util.concurrent.CompletionException: io.netty.channel.ConnectTimeoutException: connection timed out: <old.pod.ip.here>/<old.pod.ip.here>:6379
	at java.util.concurrent.CompletableFuture.encodeRelay(CompletableFuture.java:326) ~[?:1.8.0_191]
	at java.util.concurrent.CompletableFuture.completeRelay(CompletableFuture.java:338) ~[?:1.8.0_191]
	at java.util.concurrent.CompletableFuture.uniRelay(CompletableFuture.java:911) ~[?:1.8.0_191]
	at java.util.concurrent.CompletableFuture$UniRelay.tryFire(CompletableFuture.java:899) ~[?:1.8.0_191]
	... 11 more
Caused by: io.netty.channel.ConnectTimeoutException: connection timed out: <old.pod.ip.here>/<old.pod.ip.here>:6379
	at io.netty.channel.nio.AbstractNioChannel$AbstractNioUnsafe$1.run(AbstractNioChannel.java:261) ~[netty-transport-4.1.77.Final.jar:4.1.77.Final]
	at io.netty.util.concurrent.PromiseTask.runTask(PromiseTask.java:98) ~[netty-common-4.1.77.Final.jar:4.1.77.Final]
	at io.netty.util.concurrent.ScheduledFutureTask.run(ScheduledFutureTask.java:170) ~[netty-common-4.1.77.Final.jar:4.1.77.Final]
	... 8 more
```

**Expected behavior**
No error message, slave up success with new pod ip

**Steps to reproduce or test case**
create redis sentinel cluster in kubernetes with statefulset, connect java client to cluster, delete one redis slave pod in kubernetes, wait kubernetes recreate pod. see redisson client log for error message

**Redis version**
6.2.7

**Redisson version**
3.17.3

**Redisson configuration**
```

 Config config = new Config();
        try {
            SentinelServersConfig sentinelServersConfig = config.useSentinelServers()
                    .setMasterName("mymaster")
                    .addSentinelAddress("redis://" + redisURIList)
                    .setCheckSentinelsList(false)
                    .setFailedSlaveReconnectionInterval(90)
                    .setFailedSlaveCheckInterval(60)
                    .setConnectTimeout(connectionTimeout)
                    .setMasterConnectionMinimumIdleSize(connectionIdleSize)
                    .setMasterConnectionPoolSize(connectionPoolSize);
            if(userName != null && !userName.isEmpty()) {
                sentinelServersConfig.setUsername(userName);
                sentinelServersConfig.setSentinelUsername(userName);
            }
            if(password != null && !password.isEmpty()) {
                sentinelServersConfig.setPassword(new DefaultPasswordDecoder().decode(password));
                sentinelServersConfig.setSentinelPassword(new DefaultPasswordDecoder().decode(password));
            }
            return Redisson.create(config);
        } catch (Exception ex) {
            log.error("Exception while connecting to Redis ", ex);
        }
```

**Comment by mrniko** [comment]
Please share logs with `trace` logging level for `org.redisson` package.

**Comment by mingjie-li** [comment]
[trace.log](https://github.com/redisson/redisson/files/8883839/trace.log)
Here is the trace.log
we have initially 3 redis pod as below:
test-redis-0.test-redis-headless/172.17.0.4:6379 master
test-redis-1.test-redis-headless/172.17.0.2:6379 slave
test-redis-2.test-redis-headless/172.17.0.7:6379 slave

after delete the pod test-redis-2.test-redis-headless, kubernetes recreate with a new ip 172.17.0.6

then we start to see error in the trace like 

```
2022-06-11 11:47:43.847 [redisson-netty-2-1] ERROR org.redisson.connection.balancer.LoadBalancerManager - Unable to unfreeze entry: [freeSubscribeConnectionsAmount=0, freeSubscribeConnectionsCounter=value:53:queue:0, freeConnectionsAmount=0, freeConnectionsCounter=value:68:queue:0, freezeReason=MANAGER, client=[addr=redis://172.17.0.7:6379], nodeType=SLAVE, firstFail=1654948063847]
java.util.concurrent.CompletionException: org.redisson.client.RedisConnectionException: Unable to connect to Redis server: 172.17.0.7/172.17.0.7:6379
	at java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:292) ~[?:1.8.0_332]
	at java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:308) ~[?:1.8.0_332]
	at java.util.concurrent.CompletableFuture.biRelay(CompletableFuture.java:1298) ~[?:1.8.0_332]
	at java.util.concurrent.CompletableFuture$BiRelay.tryFire(CompletableFuture.java:1284) ~[?:1.8.0_332]
	at java.util.concurrent.CompletableFuture$CoCompletion.tryFire(CompletableFuture.java:1034) ~[?:1.8.0_332]
	at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:488) ~[?:1.8.0_332]
	at java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:1990) ~[?:1.8.0_332]
	at org.redisson.connection.pool.ConnectionPool.lambda$createConnection$1(ConnectionPool.java:151) ~[java-cli-1.0.0.jar:?]
	at java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:774) ~[?:1.8.0_332]
	at java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:750) ~[?:1.8.0_332]
	at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:488) ~[?:1.8.0_332]
	at java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:1990) ~[?:1.8.0_332]
	at org.redisson.connection.pool.ConnectionPool.promiseFailure(ConnectionPool.java:307) ~[java-cli-1.0.0.jar:?]
	at org.redisson.connection.pool.ConnectionPool.lambda$createConnection$6(ConnectionPool.java:273) ~[java-cli-1.0.0.jar:?]
	at java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:774) ~[?:1.8.0_332]
	at java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:750) ~[?:1.8.0_332]
	at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:488) ~[?:1.8.0_332]
	at java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:1990) ~[?:1.8.0_332]
	at org.redisson.client.RedisClient$2$2.run(RedisClient.java:292) ~[java-cli-1.0.0.jar:?]
	at io.netty.util.concurrent.AbstractEventExecutor.runTask(AbstractEventExecutor.java:174) ~[java-cli-1.0.0.jar:?]
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:167) ~[java-cli-1.0.0.jar:?]
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:470) ~[java-cli-1.0.0.jar:?]
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:503) ~[java-cli-1.0.0.jar:?]
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:995) ~[java-cli-1.0.0.jar:?]
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74) ~[java-cli-1.0.0.jar:?]
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30) ~[java-cli-1.0.0.jar:?]
	at java.lang.Thread.run(Thread.java:750) ~[?:1.8.0_332]
Caused by: org.redisson.client.RedisConnectionException: Unable to connect to Redis server: 172.17.0.7/172.17.0.7:6379
	at org.redisson.connection.pool.ConnectionPool.lambda$createConnection$1(ConnectionPool.java:150) ~[java-cli-1.0.0.jar:?]
	at java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:774) ~[?:1.8.0_332]
	at java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:750) ~[?:1.8.0_332]
	at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:488) ~[?:1.8.0_332]
	at java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:1990) ~[?:1.8.0_332]
	at org.redisson.connection.pool.ConnectionPool.promiseFailure(ConnectionPool.java:307) ~[java-cli-1.0.0.jar:?]
	at org.redisson.connection.pool.ConnectionPool.lambda$createConnection$6(ConnectionPool.java:273) ~[java-cli-1.0.0.jar:?]
	at java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:774) ~[?:1.8.0_332]
	at java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:750) ~[?:1.8.0_332]
	at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:488) ~[?:1.8.0_332]
	at java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:1990) ~[?:1.8.0_332]
	at org.redisson.client.RedisClient$1$2.run(RedisClient.java:235) ~[java-cli-1.0.0.jar:?]
	... 8 more
Caused by: java.util.concurrent.CompletionException: io.netty.channel.ConnectTimeoutException: connection timed out: 172.17.0.7/172.17.0.7:6379
	at java.util.concurrent.CompletableFuture.encodeRelay(CompletableFuture.java:326) ~[?:1.8.0_332]
	at java.util.concurrent.CompletableFuture.completeRelay(CompletableFuture.java:338) ~[?:1.8.0_332]
	at java.util.concurrent.CompletableFuture.uniRelay(CompletableFuture.java:925) ~[?:1.8.0_332]
	at java.util.concurrent.CompletableFuture$UniRelay.tryFire(CompletableFuture.java:913) ~[?:1.8.0_332]
	at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:488) ~[?:1.8.0_332]
	at java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:1990) ~[?:1.8.0_332]
	at org.redisson.client.RedisClient$1$2.run(RedisClient.java:235) ~[java-cli-1.0.0.jar:?]
	... 8 more
Caused by: io.netty.channel.ConnectTimeoutException: connection timed out: 172.17.0.7/172.17.0.7:6379
	at io.netty.channel.nio.AbstractNioChannel$AbstractNioUnsafe$1.run(AbstractNioChannel.java:261) ~[java-cli-1.0.0.jar:?]
	at io.netty.util.concurrent.PromiseTask.runTask(PromiseTask.java:98) ~[java-cli-1.0.0.jar:?]
	at io.netty.util.concurrent.ScheduledFutureTask.run(ScheduledFutureTask.java:170) ~[java-cli-1.0.0.jar:?]
	... 8 more
```
i put the kubenetes yaml for redis and java code here:
https://github.com/mingjie-li/redis

**Comment by lbornov2** [comment]
We are also facing this issue.  I can confirm that the issue happens when killing master as well.

**Comment by mrniko** [comment]
@mingjie-li 

Try to specify follow dns resolver.

```java
public class DnsAddressResolverGroupFactory implements AddressResolverGroupFactory {
    
    @Override
    public DnsAddressResolverGroup create(Class<? extends DatagramChannel> channelType,
            DnsServerAddressStreamProvider nameServerProvider) {
        DnsAddressResolverGroup group = new DnsAddressResolverGroup(new DnsNameResolverBuilder()
                .channelType(NioDatagramChannel.class)
                .ttl(1, 1)
                .nameServerProvider(DnsServerAddressStreamProviders.platformDefault());
        return group;
    }

}

config.addressResolverGroupFactory = new DnsAddressResolverGroupFactory();
```

**Comment by mingjie-li** [comment]
Hello @mrniko , we still have this issue after using your DNS class

**Comment by lbornov2** [comment]
Double that.  I still have this issue even after using the custom `DnsAddressResolverGroupFactory` provided.

**Comment by mrniko** [comment]
Please check your DNS resolver. I don't have other solution

**Comment by mrniko** [comment]
Here is another possible solution:

```java
public class DnsAddressResolverGroupFactory implements AddressResolverGroupFactory {
    
    @Override
    public DnsAddressResolverGroup create(Class<? extends DatagramChannel> channelType,
            DnsServerAddressStreamProvider nameServerProvider) {
        DnsAddressResolverGroup group = new DnsAddressResolverGroup(new DnsNameResolverBuilder()
                .channelType(NioDatagramChannel.class)
                .authoritativeDnsServerCache(NoopAuthoritativeDnsServerCache.INSTANCE)
                .nameServerProvider(DnsServerAddressStreamProviders.platformDefault());
        return group;
    }

}

config.addressResolverGroupFactory = new DnsAddressResolverGroupFactory();
```

**Comment by lbornov2** [comment]
That doesn't work either.  

Our use-case:
We run a 3-node cluster using `bitnami/redis` helm chart, version `16.9.2`.  We have 3 redis and 3 sentinel nodes.

Our relevant configuration code
```
    @Bean
    public JedisConnectionFactory redisConnectionFactory() {
        log.info("Creating clustered redis instance");
        final RedisSentinelConfiguration config = new RedisSentinelConfiguration()
                .master(sentinelMasterName);
        
        final String[] splitHost = StringUtils.split(redisHost, ",");
        final String[] splitPort = StringUtils.split(redisPort, ",");
        for(int i = 0; i < splitHost.length; i++) {
            config.sentinel(splitHost[i], Integer.valueOf(splitPort[i]));
        }
        config.setPassword(redisPassword);
        config.setSentinelPassword(sentinelPassword);
        
        JedisConnectionFactory factory = new JedisConnectionFactory(config);
        factory.getPoolConfig().setMaxTotal(maxTotal);
        factory.getPoolConfig().setMaxIdle(maxIdle);
        return factory;
    }
    
    @Bean
    public RedissonClient createRedissonClient() throws IOException {
        log.info("Creating redisson client");
        final URL url = new File(redissonYamlPath).toURI().toURL();
        final Config config = Config.fromYAML(url);
        Redisson.create(config);
    }
```

Our `redisson.yaml`:

```
---
sentinelServersConfig:
  idleConnectionTimeout: 10000
  connectTimeout: 10000
  timeout: 5000
  retryAttempts: 3
  retryInterval: 1500
  failedSlaveReconnectionInterval: 3000
  failedSlaveCheckInterval: 60000
  subscriptionsPerConnection: 5
  clientName: null
  loadBalancer: !<org.redisson.connection.balancer.RoundRobinLoadBalancer> {}
  subscriptionConnectionMinimumIdleSize: 1
  password: *****
  subscriptionConnectionPoolSize: 50
  slaveConnectionMinimumIdleSize: 24
  slaveConnectionPoolSize: 64
  masterConnectionMinimumIdleSize: 24
  masterConnectionPoolSize: 64
  readMode: "SLAVE"
  subscriptionMode: "SLAVE"
  sentinelAddresses:
  - "redis://test2021-redis-headless:26379"
  masterName: "mymaster"
  keepAlive: true
  database: 0
  checkSentinelsList: false
eventLoopGroup: null
threads: 16
nettyThreads: 32
codec: !<org.redisson.codec.SnappyCodec> {}
transportMode: "NIO"

#https://github.com/redisson/redisson/issues/4356
addressResolverGroupFactory: !<org.myapp.redis.MyDnsAddressResolverGroupFactory> {}
```

And the custom DNS Resolver that you recommended:
```

public class MyDnsAddressResolverGroupFactory implements AddressResolverGroupFactory {
    
    @Override
    public DnsAddressResolverGroup create(Class<? extends DatagramChannel> channelType,
            DnsServerAddressStreamProvider nameServerProvider) {
        DnsAddressResolverGroup group = new DnsAddressResolverGroup(new DnsNameResolverBuilder()
                .channelType(NioDatagramChannel.class)
                .authoritativeDnsServerCache(NoopAuthoritativeDnsServerCache.INSTANCE)
                .nameServerProvider(DnsServerAddressStreamProviders.platformDefault());
        return group;
    }

}
```


When killing either a master or a slave, I start seeing this in the logs immediately.  Even after the redis pod that I killed comes back up, I continue seeing this.

```
2022-06-13 12:48:00.302 ERROR 1 --- [isson-netty-2-9] o.r.c.balancer.LoadBalancerManager       : Unable to unfreeze entry: [freeSubscribeConnectionsAmount=0, freeSubscribeConnectionsCounter=value:52:queue:0, freeConnectionsAmount=0, freeConnectionsCounter=value:66:queue:0, freezeReason=MANAGER, client=[addr=redis://10.0.2.28:6379], nodeType=SLAVE, firstFail=1655124480298]

java.util.concurrent.CompletionException: org.redisson.client.RedisConnectionException: Unable to connect to Redis server: 10.0.2.28/10.0.2.28:6379
	at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:331)
	at java.base/java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:346)
	at java.base/java.util.concurrent.CompletableFuture$BiRelay.tryFire(CompletableFuture.java:1423)
	at java.base/java.util.concurrent.CompletableFuture$CoCompletion.tryFire(CompletableFuture.java:1144)
	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:506)
	at java.base/java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:2088)
	at org.redisson.connection.pool.ConnectionPool.lambda$createConnection$1(ConnectionPool.java:151)
	at java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:859)
	at java.base/java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:837)
	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:506)
	at java.base/java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:2088)
	at org.redisson.connection.pool.ConnectionPool.promiseFailure(ConnectionPool.java:307)
	at org.redisson.connection.pool.ConnectionPool.lambda$createConnection$6(ConnectionPool.java:273)
	at java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:859)
	at java.base/java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:837)
	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:506)
	at java.base/java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:2088)
	at org.redisson.client.RedisClient$2$2.run(RedisClient.java:292)
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:164)
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:469)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:503)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:986)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: org.redisson.client.RedisConnectionException: Unable to connect to Redis server: 10.0.2.28/10.0.2.28:6379
	at org.redisson.connection.pool.ConnectionPool.lambda$createConnection$1(ConnectionPool.java:150)
	at java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:859)
	at java.base/java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:837)
	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:506)
	at java.base/java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:2088)
	at org.redisson.connection.pool.ConnectionPool.promiseFailure(ConnectionPool.java:307)
	at org.redisson.connection.pool.ConnectionPool.lambda$createConnection$6(ConnectionPool.java:273)
	at java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:859)
	at java.base/java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:837)
	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:506)
	at java.base/java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:2088)
	at org.redisson.client.RedisClient$1$2.run(RedisClient.java:235)
	... 7 common frames omitted
Caused by: java.util.concurrent.CompletionException: io.netty.channel.ConnectTimeoutException: connection timed out: 10.0.2.28/10.0.2.28:6379
	at java.base/java.util.concurrent.CompletableFuture.encodeRelay(CompletableFuture.java:367)
	at java.base/java.util.concurrent.CompletableFuture.completeRelay(CompletableFuture.java:376)
	at java.base/java.util.concurrent.CompletableFuture$UniRelay.tryFire(CompletableFuture.java:1019)
	... 10 common frames omitted
Caused by: io.netty.channel.ConnectTimeoutException: connection timed out: 10.0.2.28/10.0.2.28:6379
	at io.netty.channel.nio.AbstractNioChannel$AbstractNioUnsafe$1.run(AbstractNioChannel.java:261)
	at io.netty.util.concurrent.PromiseTask.runTask(PromiseTask.java:98)
	at io.netty.util.concurrent.ScheduledFutureTask.run(ScheduledFutureTask.java:170)
	... 7 common frames omitted

2022-06-13 12:48:02.523  INFO 1 --- [isson-netty-2-6] o.r.c.SentinelConnectionManager          : sentinel: redis://10.0.2.130:26379 added
2022-06-13 12:48:03.314  WARN 1 --- [sson-netty-2-12] o.r.c.SentinelConnectionManager          : sentinel: redis://10.0.2.130:26379 is down
2022-06-13 12:48:07.525  INFO 1 --- [sson-netty-2-28] o.r.c.SentinelConnectionManager          : sentinel: redis://10.0.2.130:26379 added
2022-06-13 12:48:08.333  WARN 1 --- [sson-netty-2-30] o.r.c.SentinelConnectionManager          : sentinel: redis://10.0.2.130:26379 is down
2022-06-13 12:48:09.341  INFO 1 --- [isson-netty-2-6] o.r.c.pool.PubSubConnectionPool          : 1 connections initialized for 10.0.2.130/10.0.2.130:6379
2022-06-13 12:48:09.342  INFO 1 --- [sson-netty-2-11] o.r.c.SentinelConnectionManager          : sentinel: redis://10.0.2.130:26379 added
2022-06-13 12:48:09.352  INFO 1 --- [sson-netty-2-15] o.r.c.SentinelConnectionManager          : slave: redis://10.0.2.130:6379 added
2022-06-13 12:48:09.352  INFO 1 --- [sson-netty-2-15] o.r.connection.pool.SlaveConnectionPool  : 24 connections initialized for 10.0.2.130/10.0.2.130:6379
```


note that `10.0.2.130` is the IP of the pod which has just come up (old IP is `10.0.2.28`).

So it seems like it's hanging onto that reference somewhere.

**Comment by mingjie-li** [comment]
@lbornov2 
i put some fix here, it works for me. maybe also works for you
https://github.com/redisson/redisson/issues/4361

**Comment by lbornov2** [comment]
@mrniko - you want the netty logs from the point in time when I kill one of the (slave) redis pods?

**Comment by lbornov2** [comment]
[netty-trace.zip](https://github.com/redisson/redisson/files/8893008/netty-trace.zip)

attached is the output with `trace` for  `io.netty.resolver`, from when the redis pod was down until when it came back up.  Could only attach as a zip, sorry.  Underlying file is just a log file...

**Comment by mrniko** [comment]
@lbornov2 

Thanks for the output.

Fixed in https://github.com/redisson/redisson/issues/4361

**Comment by lbornov2** [comment]
I confirm that this is fixed by release  `3.17.4`
